### PR TITLE
ENH PHP 8.1 compatibility

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/cow/issues/232

**Changes:**
_note: the checkboxes are so that the reviewer can more easily say what they've checked since changes aren't separated as separate commits_

- [ ] Require PHP 8.1
- [x] Update dependencies
- [ ] Adding TEST_MODE which prevents pushing to github and transifex
- [x] No longer use -release branches
- [ ] Create minor branches if required for beta and rc tags (though still not for alpha)
- [ ] Add in graphql hack
- [x] Use HEAD for "to-version" in audit changelog
- [x] Show post-rc commits in audit changelog

Requires https://github.com/silverstripe/gha-generate-matrix/pull/42 to be merged and released for unit tests to pass

### CODE REVIEW

**Comment by GuySartorelli** [comment]
CI still failing after releasing the gha-generate-matrix change and rerunning all jobs

**Comment by GuySartorelli** [comment]
prf-low and phplinting jobs failed

**Comment by GuySartorelli** [comment]
Next time you do a bunch of unrelated changes in a single PR like this _please please please_ add them as separate commits! It would make reviewing this _way_ less of a headache if I could review each piece of functionality at a time without having to figure out what's what.

**Comment by GuySartorelli** [comment]
The stabalisation changes aren't mentioned in the description??

**Comment by emteknetnz** [comment]
The stabilisation changes are covered by "No longer use -release branches" - I'm assuming the stabilisation was working correctly beforehand.

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [inline src/Steps/Release/MergeUpRelease.php:87]: Is the `$isMajorBranch` parameter for `mergeUp()` needed anymore? I don't think it is?

File: src/Model/Release/LibraryRelease.php
Comment (GuySartorelli) [inline src/Model/Release/LibraryRelease.php:350]: We should probably implement this as `isOnMinorReleaseBranch()` or similarly named instead (i.e. it's on _the correct_ minor branch) and use the plan to determine the correct branch.

Though if you think that's more effort than it's worth I will defer to your judgment.

File: src/Model/Release/ComposerConstraint.php
Comment (GuySartorelli) [inline src/Model/Release/ComposerConstraint.php:24]: I feel like it would be too easy to forget to update this if we release a new patch.
How important is that `.1` there as opposed to `.0`? If we're going to specify the exact patch version I think we should get that from the github api instead of hardcoding it.

File: src/Model/Changelog/Changelog.php
Comment (GuySartorelli) [inline src/Model/Changelog/Changelog.php:132]: This change will fail linting

File: src/Model/Changelog/Changelog.php
Comment (GuySartorelli) [inline src/Model/Changelog/Changelog.php:53]: Why make this a tri-state instead of just default to `false`?

File: src/Commands/Command.php
Comment (GuySartorelli) [inline src/Commands/Command.php:67]: We should return one of the constants defined on `Command` - and for that matter we should probably just get `fire()` to return the appropriate const based on whether it was successful or not.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Commands/Command.php
Comment (emteknetnz) [reply src/Commands/Command.php:67]: Have changed the default `1` to Console\Command\Command::SUCCESS

I don't think updating the 17 fire() methods is worth it, if we did this we'd also want to return `Console\Comman\Command::FAILURE` everytime there's some sort of error, so we'd need to update lots of code.

I really don't think slightly nicer output colours (I'm presuming that's what the outcome) is worth the effort.

Comment (GuySartorelli) [reply src/Commands/Command.php:67]: Yeah, fair enough. I think cow is a mangled enough mess that the benefits aren't really beneficial without a considerable refactor.

File: src/Model/Changelog/Changelog.php
Comment (emteknetnz) [reply src/Model/Changelog/Changelog.php:53]: Have changed to regular `bool`

File: src/Model/Changelog/Changelog.php
Comment (emteknetnz) [reply src/Model/Changelog/Changelog.php:132]: Updated

File: src/Model/Release/ComposerConstraint.php
Comment (emteknetnz) [reply src/Model/Release/ComposerConstraint.php:24]: In practice it makes no difference.

Using the github api from ComposerConstraint would be nicer, but it requires a lot of code change to do this.
There's actually 2x github api clients
- GithubApi (old, seems unused)
- PublishRelease::getGithubClient() (the one that is actually used e.g. creating github releases)

Unfortunately PublishRelease::getGithubClient() cannot easily be separated (I tried) due to it needing to instantiate a StepCommandRunner, which itself needs a Step.  It's basically just needs a solid refactor in order to use it.

The chance of graphql 3.9 being released at this stage seems fairly remote.  And if it is, it'll be in the context of a security release, and I'd expect we'd be pretty conscious of checking the what's released in the CMS 4 version of recipe-cms at that stage.

Comment (GuySartorelli) [reply src/Model/Release/ComposerConstraint.php:24]: > In practice it makes no little difference if it's .1 or .0. It would only affect a --prefer-lowest CI build.

Let's set it to `.0` then to make the intention clearer (i.e. we don't care that patch version is, what we care about is that it's stable)

File: src/Model/Release/ComposerConstraint.php
Comment (emteknetnz) [reply src/Model/Release/ComposerConstraint.php:143]: Only when it's .x-dev format. I've run this through an entire simlulated (DEV_MODE=1) release for 4.13.0, the tagged versions were what was expected for at each point.

File: src/Model/Release/LibraryRelease.php
Comment (emteknetnz) [reply src/Model/Release/LibraryRelease.php:350]: I don't think it's worth it at this stage, I've run simulated (DEV_MODE=1) releases for 4.13.0 and 5.0.0-beta1 and thing seemed to work as they should.

File: src/Steps/Release/MergeUpRelease.php
Comment (emteknetnz) [reply src/Steps/Release/MergeUpRelease.php:87]: Yes, it's passed to ConstraintDestabiliser::destabiliseConstraints() which does use it

File: src/Model/Release/LibraryRelease.php
Comment (GuySartorelli) [reply src/Model/Release/LibraryRelease.php:350]: This suggestion is more to protect against _human_ error (e.g. I manually checkout `4.12` after resolving merge conflicts, but I should have typed `4.13`)

File: src/Steps/Release/MergeUpRelease.php
Comment (GuySartorelli) [reply src/Steps/Release/MergeUpRelease.php:87]: Right, but that's just always going to be true for mergup, right? So we can remove the parameter from mergeup and just pass in true from there.

File: src/Steps/Release/RewriteReleaseBranches.php
Comment (emteknetnz) [reply src/Steps/Release/RewriteReleaseBranches.php:261]: Because new branches weren't being created for beta releases when the default "auto" branches was used.  Saves having to change the branching strategy to "minor" in the plan

Since we always start with a beta (we didn't used to) this change makes sense to me

File: src/Steps/Step.php
Comment (emteknetnz) [reply src/Steps/Step.php:105]: `Symfony\Component\Process\ProcessBuilder` was removed. Replacement was simply to use `new Process($command);`

File: src/Steps/Step.php
Comment (emteknetnz) [reply src/Steps/Step.php:120]: `[Symfony 5] The "Symfony\Component\Process\Process::inheritEnvironmentVariables()" method is deprecated since Symfony 4.4, env variables are always inherited`

Removed and no need for a replacement

File: src/Steps/Release/RewriteReleaseBranches.php
Comment (GuySartorelli) [reply src/Steps/Release/RewriteReleaseBranches.php:261]: Ahhh okay that's what the branching strategy does. I assume this will come with a relevant change in the process doc as well.
Sounds good.

File: src/Steps/Release/PublishRelease.php
Comment (emteknetnz) [reply src/Steps/Release/PublishRelease.php:96]: Updated

File: src/Steps/Release/PublishRelease.php
Comment (emteknetnz) [reply src/Steps/Release/PublishRelease.php:144]: Updated

File: src/Steps/Release/RewriteReleaseBranches.php
Comment (emteknetnz) [reply src/Steps/Release/RewriteReleaseBranches.php:261]: > Ahhh okay that's what the branching strategy does

I honestly don't know what it does and whatever it does it's probably busted.  I don't even know what I'd document.

File: src/Model/Release/LibraryRelease.php
Comment (emteknetnz) [reply src/Model/Release/LibraryRelease.php:350]: Updated

File: src/Steps/Release/RewriteReleaseBranches.php
Comment (GuySartorelli) [reply src/Steps/Release/RewriteReleaseBranches.php:261]: Oh okay. I thought you'd figured it out.
No worries then, so long as it's clear from the confluence doc what to do, I'm not fussed.

File: src/Model/Release/ComposerConstraint.php
Comment (emteknetnz) [reply src/Model/Release/ComposerConstraint.php:24]: `.1` is better, it could make a difference on `--prefer-lowest` builds for CMS 4

When I say "in practice" - no-one other than our CI use --prefer-lowest.  It's still nice to have for our own purposes.

Comment (GuySartorelli) [reply src/Model/Release/ComposerConstraint.php]: Then I revert back to my original position. If we care about the patch version we should just the API to get the correct patch version, since we're not likely to remember to update that as we do patch releases.

File: src/Steps/Release/MergeUpRelease.php
Comment (emteknetnz) [reply src/Steps/Release/MergeUpRelease.php:87]: mergeUp() is currently agnostic if it's merging minor->minor or minor->major.

I know that our current workflow is always minor->major, but they may change in the future so may as well just leave it agnostic.  The recommended change would force it to be minor->major (or major->major) ... however all the language and variable names within mergeUp() are written for it being agnostic

tl;dr just leave it as is

File: src/Model/Release/ComposerConstraint.php
Comment (emteknetnz) [reply src/Model/Release/ComposerConstraint.php]: As mentioned above I think there's too much refactoring required at this stage to get this working.  We can open a new card if you think it's worthwhile to look at later

Comment (GuySartorelli) [reply src/Model/Release/ComposerConstraint.php]: This hasn't changed since last time it was assigned to me.... I'll just let it stay as is, it's not a big deal.

File: src/Model/Release/LibraryRelease.php
Comment (GuySartorelli) [inline src/Model/Release/LibraryRelease.php:352]: `$this->version` is what's in the plan, right? Won't this then always return false if we're doing a beta or rc (where version has `-beta1` at the end for example).
If so, is that intentional? I'm trying to understand how this condition is helping us.

Comment (emteknetnz) [reply src/Model/Release/LibraryRelease.php]: Ah yep, have updated to remove the trailing '$' so it'll match beta1 etc

Main thing is to extra the major.minor so we get the expected branch name

Comment (GuySartorelli) [reply src/Model/Release/LibraryRelease.php]: Sweet. I don't think the condition is needed but it also doesn't hurt anything so I'm okay with it staying for this internal thing.

**Review by GuySartorelli — APPROVED** [review]

**Pull request merged**
